benchmark/benchmark.py

```diff
@@ -0,0 +1,84 @@
+# Benchmarking utility for PennyLane
+import functools
+import importlib
+import subprocess
+import sys
+
+import pennylane as qml
+
+
+def timing(func):
+    "Time the given function."
+    import timeit
+
+    res = timeit.repeat(func, number=10, repeat=5, globals=globals())
+    print("Timing:", res)
+
+
+def plot(title, kernels, n_range):
+    "Plot timings as a function of the parameter n."
+    import perfplot
+
+    labels = [k.__doc__ for k in kernels]
+    perfplot.show(
+        setup=lambda n: n,
+        kernels=kernels,
+        labels=labels,
+        n_range=n_range,
+        xlabel="n",
+        title=title,
+        # logx="auto",  # set to True or False to force scaling
+        # logy="auto",
+        # equality_check=numpy.allclose,  # set to None to disable "correctness" assertion
+        # automatic_order=True,
+        # colors=None,
+        # target_time_per_measurement=1.0,
+        # time_unit="s",  # set to one of ("auto", "s", "ms", "us", or "ns") to force plot units
+        # relative_to=1,  # plot the timings relative to one of the measurements
+        # flops=lambda n: 3*n,  # FLOPS plots
+    )
+
+
+def profile(func):
+    "Profile the given function."
+    import cProfile, pstats
+
+    pr = cProfile.Profile()
+    pr.enable()
+
+    for k in range(20):
+        func()
+
+    pr.disable()
+    pr.dump_stats("pennylane_{}.pstats".format(qml.version()))
+    ps = pstats.Stats(pr).strip_dirs().sort_stats("tottime")  # "cumulative")
+    ps.print_stats()
+
+
+if __name__ == "__main__":
+    # import pdb; pdb.set_trace()
+
+    n = len(sys.argv)
+    if n != 3:
+        print("Usage: benchmark.py {time,plot,profile} benchmark_module")
+        sys.exit(0)
+
+    res = subprocess.run(["git", "log", "-1", "--pretty=%h %s"], stdout=subprocess.PIPE)
```

Comment:
This line here pulls the short-form hash and subject of the current Git commit.

Reply:
Uh, I just realized that keeping the benchmarking framework in the same repository as the code is not a good idea when you want to compare the performance of different commits, since when you check out a commit, the benchmark script will change also, or not even exist at that point in time. Any good ideas?

Reply:
I guess the only solution is a separate benchmarking repo? :laughing: Or instructions to make a copy of the benchmarking script.

Reply:
Copying the `benchmark` directory to a different one works and is pretty easy, I guess that could be the solution for now.

Reply:
How about incorporating the benchmarks as profiling for the PR? 

When benchmarking locally, there could potentially other processes making use of the resources, but when there is a clean virtual machine running a new code, that potentially it would be easier to.

The easiest way would be running the benchmark script in Travis after the test suite. The more complicated way would be trying to fire up a third party service that also posts here the results. This would, however, require extra work for sure, it is mostly a long-shot idea. :)

Reply:
> How about incorporating the benchmarks as profiling for the PR using Travis and/or a third party?

This is a great idea. Another option is using [GitHub actions](https://github.com/actions)

Reply:
That might not be as useful as one might think. Isn't the virtual machine running on a CPU with other processes, so their performance is not predictable or constant? There's also the question which benchmark to run, we'd first have to consider what kind of benchmarks are the most relevant ones. And whenever you change a benchmark it makes all the old results non-comparable to the new ones.

Reply:
1. So in our travis script we have ` dist: xenial`. Looking at [the description from Travis](https://docs.travis-ci.com/user/reference/overview/#virtualisation-environment-vs-operating-system), this indicates that we have "Virtual machine on GCE". Here, GCE will stand for "Google Compute Engine", which according to [Wiki](https://en.wikipedia.org/wiki/Google_Compute_Engine#cite_note-8): "Google Compute Engine enables users to launch virtual machines (VMs) on demand." So it is also described that this is Google's Infrastructure as a service (IaaS) solution. Meaning that once we are starting up our Travis scripts, there is a request going to GCE for available resources (as detailed on the Travis website) and a virtual machine will be created, once such resources were found to be available. Now though it is true that a single server would be running multiple VMs like this, they are isolated from each other. Although these VMs are united by one hypervisor that is basically an overlay entity creating the VMs and managing the communication between the server's operating system and the VMs, the HW resources are distributed according to their original allocation. What this means, is that each VM essentially acts as a separate machine.

That being said, since a VM is fired up each time allocating the same resources every time, assuming that there is no problem on the third party's side (GCE), I would say it is a really good environment to run benchmarks in. The reason is, is that the processes running our scripts would be the only ones running in those VMs.

2. I agree! For now, maybe we could go with anything that we deem would be necessary for benchmarking performance, but altogether it would indeed be really nice to have a set list. Also benchmarking itself should not take too long overall (e.g. not more than 5 minutes, only if really necessary, but such benchmarks could also be organized separately).

3. That's a fair point, however, there will be *evergreen* benchmarks that will be expected to perform most of the times similarly (such as running the single-qubit operations in a `QNode`). These would not be affected since we would always want to have `QNodes` that are running only one one-qubit operation on "standardized" devices (same configurations such as simulator, number of qubits, number of shots, analytic values, etc.). Other benchmark cases could cover devices from each of the plugins.

When a benchmark is changed such that its new results are not easy to be interpreted, they could just be regarded as new benchmarks (meaning that for lack of prior data we would not be able to come to any conclusions just yet).

Overall, for benchmarking, as we are depending on the speed of many other packages as well, one of the goals here could be to **quickly detect** if there was an *unexpected* deviation in speed. An expected deviation would be for example adding extra validation or extra checks to the `QNode` and then experiencing a basic quantum circuit running 5% slower.

Reply:
I'm not sure that we can guarantee that Travis will always provision VMs with the same resources, however. This could cause stochastic failures, lowering confidence in the benchmarking results.

Something similar occurs on readthedocs - VMs are provisioned from different pools, and the VM you are provided with can greatly affect whether the tutorials build/the speed they build at.

We would have to look into it, but I think a subscription is required in order to specify VM resources.

For benchmarking, I suggest this is not a test run with every PR, but instead one we run on a schedule.

Reply:
There's also the option of running the benchmarks manually, but having a script that automatically checks out a predetermined list of commits (tagged releases for example) and benchmarks each one of them, and then e.g. produces a plot that shows how the performance has changed over the releases.

Reply:
Let's continue this discussion once this PR has been merged. Perhaps the best approach is to run the benchmark script internally on a system we can control (e.g., Jenkins), including the comparison between tagged commits as suggested above.